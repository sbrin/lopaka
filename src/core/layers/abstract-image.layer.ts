import {TPlatformFeatures} from '../../platforms/platform';
import {
    downloadImage,
    flipImageDataByX,
    flipImageDataByY,
    hexToRgb,
    invertImageData,
    rotateImageData
} from '../../utils';
import {mapping} from '../decorators/mapping';
import {Point} from '../point';
import {Rect} from '../rect';
import {AbstractLayer, EditMode, TLayerActions} from './abstract.layer';

export abstract class AbstractImageLayer extends AbstractLayer {
    protected type: ELayerType;
    protected editState: {
        firstPoint: Point;
        position: Point;
        size: Point;
    } = null;

    public get properties(): any {
        return {
            x: this.position.x,
            y: this.position.y,
            w: this.size.x,
            h: this.size.y,
            overlay: this.overlay,
            color: this.color,
            image: this.data,
            type: this.type,
            id: this.uid,
            imageName: this.imageName,
            inverted: this.inverted
        };
    }

    @mapping('p', 'point') public position: Point = new Point();

    @mapping('s', 'point') public size: Point = new Point();

    @mapping('d', 'image') public data: ImageData;

    @mapping('o') public overlay: boolean;

    @mapping('nm') public imageName: string;

    constructor(protected features: TPlatformFeatures) {
        super(features);
    }

    actions: TLayerActions = [
        {
            label: '⬌',
            title: 'Flip horizontally',
            action: () => {
                this.data = flipImageDataByX(this.data);
                this.draw();
            }
        },
        {
            label: '⬍',
            title: 'Flip vertically',
            action: () => {
                this.data = flipImageDataByY(this.data);
                this.draw();
            }
        },
        {
            label: '↻',
            title: 'Rotate clockwise',
            action: () => {
                this.data = rotateImageData(this.data);
                this.size = new Point(this.data.width, this.data.height);
                this.updateBounds();
                this.draw();
            }
        },
        {
            label: '◧',
            title: 'Invert colors',
            action: () => {
                this.data = invertImageData(this.data, this.color);
                this.draw();
            }
        },
        {
            label: '⚀',
            title: 'Remove blank padding',
            action: () => {
                this.recalculate();
                this.updateBounds();
                this.draw();
            }
        },
        {
            label: 'Download',
            title: 'Download image',
            action: () => {
                downloadImage(this.data, this.imageName ?? 'image_' + this.index);
            }
        }
    ];

    applyColor() {
        const color = hexToRgb(this.color);
        this.data.data.forEach((v, i) => {
            if (i % 4 === 3 && v !== 0) {
                this.data.data[i - 3] = color.r;
                this.data.data[i - 2] = color.g;
                this.data.data[i - 1] = color.b;
                return v;
            }
        });
    }

    recalculate() {
        const {dc} = this;
        const {width, height} = dc.ctx.canvas;
        const data = dc.ctx.getImageData(0, 0, width, height);
        let min: Point = new Point(width, height);
        let max: Point = new Point(0, 0);
        dc.ctx.beginPath();
        for (let i = 0; i < data.data.length; i += 4) {
            const x = (i / 4) % width;
            const y = Math.floor(i / 4 / width);
            if (data.data[i + 3] !== 0) {
                min = min.min(new Point(x, y));
                max = max.max(new Point(x, y));
            }
        }
        this.position = min.clone();
        this.size = max.clone().subtract(min).add(1);
        this.data = dc.ctx.getImageData(min.x, min.y, this.size.x, this.size.y);
    }

    draw() {
        const {dc, position, data, size} = this;
        dc.clear();
        dc.ctx.putImageData(data, position.x, position.y);
        dc.ctx.save();
        dc.ctx.fillStyle = 'rgba(0,0,0,0)';
        dc.ctx.beginPath();
        dc.ctx.rect(position.x, position.y, size.x, size.y);
        dc.ctx.fill();
        dc.ctx.restore();
    }

    onLoadState() {
        this.updateBounds();
        this.mode = EditMode.NONE;
        this.applyColor();
    }

    updateBounds(): void {
        this.bounds = new Rect(this.position, this.size);
    }
}
