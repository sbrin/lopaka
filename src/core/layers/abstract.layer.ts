import {DrawContext} from '../../draw/draw-context';
import {TPlatformFeatures} from '../../platforms/platform';
import {generateUID} from '../../utils';
import {Point} from '../point';
import {Rect} from '../rect';
// TODO move type delarations outside of the class
export enum EditMode {
    MOVING,
    RESIZING,
    CREATING,
    DRAWING,
    NONE,
    EMPTY
}
export enum TModifierType {
    'string',
    'number',
    'boolean',
    'font',
    'image',
    'color'
}
export type TModifierName =
    | 'x'
    | 'y'
    | 'w'
    | 'h'
    | 'radius'
    | 'text'
    | 'font'
    | 'icon'
    | 'x1'
    | 'x2'
    | 'y1'
    | 'y2'
    | 'color'
    | 'image'
    | 'overlay'
    | 'color'
    | 'fontSize';

export type TLayerModifier = {
    setValue(value: any): void;
    getValue(): any;
    type: TModifierType;
};

export type TLayerModifiers = Partial<{[key in TModifierName]: TLayerModifier}>;

export type TLayerEditPoint = {
    cursor: 'nwse-resize' | 'nesw-resize' | 'move';
    getRect(): Rect;
    move(point: Point): void;
};

export type TLayerState = {
    t: ELayerType; // type
    n: string; // name
    i: number; // index
    g: number; // group
    u: string; // uid
    c?: string; // color
};

/**
 * Abstract layer class
 */
export abstract class AbstractLayer {
    protected abstract type: ELayerType;
    // OffscreenCanvas where layer is drawn
    protected buffer: OffscreenCanvas = new OffscreenCanvas(0, 0);
    // DrawContext
    protected dc: DrawContext = new DrawContext(this.buffer);
    // Unique id
    protected id: string = '';
    // Is layer visible
    protected isOverlay: boolean = false;
    // current layer state
    protected state: any = {};
    // current edit mode
    protected mode: EditMode = EditMode.EMPTY;
    // history of changing
    protected history: any[] = [];
    // UID
    public uid = generateUID();

    // is layer selected
    public selected: boolean = false;
    // Bounds of the layer
    public bounds: Rect = new Rect();
    // Layer name
    public name: string;
    // Layer index
    public index: number;
    // public group
    public group: number;
    // is layer already added to the session
    public added: boolean = false;
    // is layer resizable
    public resizable: boolean = true;
    // modifiers
    public modifiers: TLayerModifiers = {};
    // color
    public color: string = '#000000';

    public editPoints: TLayerEditPoint[] = [];

    constructor(protected features?: TPlatformFeatures) {}

    // called when layer starts to edit
    abstract startEdit(mode: EditMode, point?: Point, editPoint?: TLayerEditPoint);
    // called when layer is editing
    abstract edit(point: Point, originalEvent?: MouseEvent);
    // called when layer stops to edit
    abstract stopEdit();
    // draw layer
    abstract draw();
    // save layer state
    abstract saveState();
    // load layer state
    abstract loadState(state: any);
    // update layer bounds
    abstract updateBounds(): void;

    /**
     * Check that point inside layer
     * @param point
     * @returns {boolean}
     */
    public contains(point: Point): boolean {
        return this.dc.ctx.isPointInPath(point.x, point.y) || this.dc.ctx.isPointInStroke(point.x, point.y);
    }

    /**
     * Resize buffer to fit display
     * @param display
     * @param scale
     */
    public resize(display: Point, scale: Point): void {
        const {dc, buffer} = this;
        buffer.width = display.x;
        buffer.height = display.y;
        dc.ctx.fillStyle = '#000';
        dc.ctx.strokeStyle = '#000';
        if (this.mode !== EditMode.EMPTY) {
            this.draw();
        }
    }

    /**
     * Get layer buffer
     * @returns {OffscreenCanvas} - buffer
     */
    public getBuffer(): OffscreenCanvas {
        return this.buffer;
    }

    /**
     * Clone this layer as new one
     * @returns copy of the layer
     */
    public clone(): typeof this {
        const cloned = new (this.constructor as any)();
        cloned.loadState(this.state);
        return cloned;
    }

    /**
     * Is layer in resizing mode
     * @param state
     */
    public isResizing(): boolean {
        return this.mode === EditMode.RESIZING;
    }

    /**
     * Is layer in editing mode
     * @returns {boolean}
     */
    public isEditing() {
        return this.mode !== EditMode.NONE;
    }

    public getState(): any {
        return this.state;
    }

    public getType() {
        return this.type;
    }
}
