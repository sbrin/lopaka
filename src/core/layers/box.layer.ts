import {RectangleLayer} from './rectangle.layer';

export class BoxLayer extends RectangleLayer {
    protected type: ELayerType = 'box';
    protected fill: true;
}
