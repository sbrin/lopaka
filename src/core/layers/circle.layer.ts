import {TPlatformFeatures} from '../../platforms/platform';
import {Point} from '../point';
import {Rect} from '../rect';
import {AbstractLayer, EditMode, TLayerEditPoint, TLayerModifiers, TLayerState, TModifierType} from './abstract.layer';

type TCircleState = TLayerState & {
    p: number[]; // position
    r: number; // radius
};

export class CircleLayer extends AbstractLayer {
    protected type: ELayerType = 'circle';
    protected state: TCircleState;
    public radius: number = 1;
    public position: Point = new Point();
    protected fill: boolean = false;

    protected editState: {
        firstPoint: Point;
        position: Point;
        radius: number;
        editPoint: TLayerEditPoint;
    } = null;

    modifiers: TLayerModifiers = {
        x: {
            getValue: () => this.position.x,
            setValue: (v: number) => {
                this.position.x = v;
                this.updateBounds();
                this.saveState();
                this.draw();
            },
            type: TModifierType.number
        },
        y: {
            getValue: () => this.position.y,
            setValue: (v: number) => {
                this.position.y = v;
                this.updateBounds();
                this.saveState();
                this.draw();
            },
            type: TModifierType.number
        },
        radius: {
            getValue: () => this.radius,
            setValue: (v: number) => {
                this.radius = v;
                this.updateBounds();
                this.saveState();
                this.draw();
            },
            type: TModifierType.number
        },
        color: {
            getValue: () => this.color,
            setValue: (v: string) => {
                this.color = v;
                this.updateBounds();
                this.saveState();
                this.draw();
            },
            type: TModifierType.color
        }
    };

    constructor(protected features: TPlatformFeatures) {
        super(features);
        if (!this.features.hasCustomFontSize) {
            delete this.modifiers.fontSize;
        }
        if (!this.features.hasRGBSupport) {
            delete this.modifiers.color;
            this.color = '#000000';
        }
        if (this.features.hasInvertedColors) {
            this.color = '#FFFFFF';
        }
    }

    editPoints: TLayerEditPoint[] = [
        {
            cursor: 'nesw-resize',
            getRect: (): Rect =>
                new Rect(new Point(this.bounds.x + this.bounds.w, this.bounds.y), new Point(3)).subtract(
                    1.5,
                    1.5,
                    0,
                    0
                ),
            move: (offset: Point): void => {
                this.position.y = Math.floor(this.editState.position.y + Math.round(offset.x / 2) * 2);
                this.radius = Math.floor(this.editState.radius - offset.x / 2);
            }
        },
        {
            cursor: 'nwse-resize',
            getRect: (): Rect =>
                new Rect(
                    new Point(this.bounds.x + this.bounds.w, this.bounds.y + this.bounds.h),
                    new Point(3)
                ).subtract(1.5, 1.5, 0, 0),
            move: (offset: Point): void => {
                this.radius = Math.round(this.editState.radius - offset.x / 2);
            }
        },
        {
            cursor: 'nesw-resize',
            getRect: (): Rect =>
                new Rect(new Point(this.bounds.x, this.bounds.y + this.bounds.h), new Point(3)).subtract(
                    1.5,
                    1.5,
                    0,
                    0
                ),
            move: (offset: Point): void => {
                this.position.x = Math.ceil(this.editState.position.x - Math.round(offset.x / 2) * 2);
                this.radius = Math.ceil(this.editState.radius + offset.x / 2);
            }
        },
        {
            cursor: 'nwse-resize',
            getRect: (): Rect =>
                new Rect(new Point(this.bounds.x, this.bounds.y), new Point(3)).subtract(1.5, 1.5, 0, 0),
            move: (offset: Point): void => {
                this.position = this.editState.position
                    .clone()
                    .subtract(Math.round(offset.x / 2) * 2)
                    .ceil();
                this.radius = Math.ceil(this.editState.radius + offset.x / 2);
            }
        }
    ];

    startEdit(mode: EditMode, point: Point, editPoint: TLayerEditPoint) {
        this.mode = mode;
        if (mode == EditMode.CREATING) {
            this.position = point.clone();
            this.radius = 1;
            this.updateBounds();
            this.draw();
        }
        this.editState = {
            firstPoint: point,
            position: this.position.clone(),
            radius: this.radius,
            editPoint
        };
    }

    edit(point: Point, originalEvent: MouseEvent) {
        const {position, firstPoint, editPoint} = this.editState;
        switch (this.mode) {
            case EditMode.MOVING:
                this.position = position.clone().add(point.clone().subtract(firstPoint)).round();
                break;
            case EditMode.RESIZING:
                // todo
                editPoint.move(firstPoint.clone().subtract(point));
                break;
            case EditMode.CREATING:
                const radius = Math.max(
                    Math.max(...point.clone().subtract(firstPoint).abs().divide(2).round().xy) - 2,
                    1
                );
                this.radius = radius;
                if (originalEvent.altKey) {
                    this.position = firstPoint.clone().subtract(radius);
                } else {
                    const signs = point.clone().subtract(firstPoint).xy.map(Math.sign);
                    this.position = firstPoint.min(firstPoint.clone().add(new Point(radius * 2 + 1).multiply(signs)));
                }
                break;
        }
        this.updateBounds();
        this.draw();
    }

    stopEdit() {
        this.mode = EditMode.NONE;
        this.editState = null;
        this.saveState();
        this.history.push(this.state);
    }

    draw() {
        const {dc, radius, position} = this;
        const center = position.clone().add(radius);
        dc.clear();
        dc.ctx.fillStyle = this.color;
        dc.ctx.strokeStyle = this.color;
        dc.pixelateCircle(center, radius, this.fill);
    }

    saveState() {
        const state: TCircleState = {
            p: this.position.xy,
            r: this.radius,
            n: this.name,
            i: this.index,
            g: this.group,
            t: this.type,
            u: this.uid,
            c: this.color
        };
        this.state = state;
    }

    loadState(state: TCircleState) {
        this.position = new Point(state.p);
        this.radius = state.r;
        this.name = state.n;
        this.index = state.i;
        this.group = state.g;
        this.uid = state.u;
        this.color = state.c;
        this.updateBounds();
    }

    updateBounds() {
        this.bounds = new Rect(this.position, new Point(this.radius * 2)).add(0, 0, 1, 1);
    }
}
