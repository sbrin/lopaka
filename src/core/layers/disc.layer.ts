import {CircleLayer} from './circle.layer';

export class DiscLayer extends CircleLayer {
    protected type: ELayerType = 'disc';
    protected fill: boolean = true;
}
