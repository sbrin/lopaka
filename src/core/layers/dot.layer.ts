import {TPlatformFeatures} from '../../platforms/platform';
import {Point} from '../point';
import {Rect} from '../rect';
import {EditMode} from './abstract.layer';
import {RectangleLayer} from './rectangle.layer';

export class DotLayer extends RectangleLayer {
    protected type: ELayerType = 'dot';
    protected brushSize: Point = new Point(1);
    public size: Point = this.brushSize.clone();
    public resizable: boolean = false;
    protected fill: boolean = true;
    constructor(protected features: TPlatformFeatures) {
        super(features);
        delete this.modifiers.w;
        delete this.modifiers.h;
        if (!this.features.hasCustomFontSize) {
            delete this.modifiers.fontSize;
        }
        if (!this.features.hasRGBSupport) {
            delete this.modifiers.color;
            this.color = '#000000';
        }
        if (this.features.hasInvertedColors) {
            this.color = '#FFFFFF';
        }
    }

    edit(point: Point, originalEvent: MouseEvent) {
        switch (this.mode) {
            case EditMode.MOVING:
            case EditMode.CREATING:
                this.position = point.clone();
                break;
        }
        this.bounds = new Rect(this.position, this.size);
        this.draw();
    }
}
