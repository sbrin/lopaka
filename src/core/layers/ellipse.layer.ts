import {TPlatformFeatures} from '../../platforms/platform';
import {Point} from '../point';
import {Rect} from '../rect';
import {AbstractLayer, EditMode, TLayerEditPoint, TLayerModifiers, TLayerState, TModifierType} from './abstract.layer';

type TEllipseState = TLayerState & {
    p: number[]; // position
    rx: number; // radius X
    ry: number; // radius Y
    f: boolean; // fill
};

export class EllipseLayer extends AbstractLayer {
    protected type: ELayerType = 'ellipse';
    protected state: TEllipseState;
    public radiusX: number = 1;
    public radiusY: number = 1;
    public position: Point = new Point();
    public fill: boolean = false;

    protected editState: {
        firstPoint: Point;
        position: Point;
        radiusX: number;
        radiusY: number;
        editPoint: TLayerEditPoint;
    } = null;

    public get properties(): any {
        return {
            x: this.position.x + this.radiusX,
            y: this.position.y + this.radiusY,
            rx: this.radiusX,
            ry: this.radiusY,
            fill: this.fill,
            color: this.color,
            type: this.type,
            id: this.uid,
            inverted: this.inverted
        };
    }

    modifiers: TLayerModifiers = {
        x: {
            getValue: () => this.position.x,
            setValue: (v: number) => {
                this.position.x = v;
                this.updateBounds();
                this.saveState();
                this.draw();
            },
            type: TModifierType.number
        },
        y: {
            getValue: () => this.position.y,
            setValue: (v: number) => {
                this.position.y = v;
                this.updateBounds();
                this.saveState();
                this.draw();
            },
            type: TModifierType.number
        },
        radiusX: {
            getValue: () => this.radiusX,
            setValue: (v: number) => {
                this.radiusX = v;
                this.updateBounds();
                this.saveState();
                this.draw();
            },
            type: TModifierType.number
        },
        radiusY: {
            getValue: () => this.radiusY,
            setValue: (v: number) => {
                this.radiusY = v;
                this.updateBounds();
                this.saveState();
                this.draw();
            },
            type: TModifierType.number
        },
        fill: {
            getValue: () => this.fill,
            setValue: (v: boolean) => {
                this.fill = v;
                this.saveState();
                this.draw();
            },
            type: TModifierType.boolean
        },
        color: {
            getValue: () => this.color,
            setValue: (v: string) => {
                this.color = v;
                this.updateBounds();
                this.saveState();
                this.draw();
            },
            type: TModifierType.color
        },
        inverted: {
            getValue: () => this.inverted,
            setValue: (v: boolean) => {
                this.inverted = v;
                this.saveState();
                this.draw();
            },
            type: TModifierType.boolean
        }
    };

    constructor(protected features: TPlatformFeatures) {
        super(features);
        if (!this.features.hasRGBSupport) {
            delete this.modifiers.color;
        }
        if (!this.features.hasInvertedColors) {
            delete this.modifiers.inverted;
        }
        this.color = this.features.defaultColor;
    }

    editPoints: TLayerEditPoint[] = [
        {
            cursor: 'nesw-resize',
            getRect: (): Rect =>
                new Rect(new Point(this.bounds.x + this.bounds.w, this.bounds.y), new Point(3)).subtract(
                    1.5,
                    1.5,
                    0,
                    0
                ),
            move: (offset: Point): void => {
                const dx = Math.round(offset.x / 2);
                const dy = Math.round(offset.y / 2);
                if (Math.floor(this.editState.radiusY + dy) >= 1) {
                    this.position.y = this.editState.position.y - dy * 2;
                }
                this.radiusX = Math.max(1, Math.floor(this.editState.radiusX - dx));
                this.radiusY = Math.max(1, Math.floor(this.editState.radiusY + dy));
            }
        },
        {
            cursor: 'nwse-resize',
            getRect: (): Rect =>
                new Rect(
                    new Point(this.bounds.x + this.bounds.w, this.bounds.y + this.bounds.h),
                    new Point(3)
                ).subtract(1.5, 1.5, 0, 0),
            move: (offset: Point): void => {
                const dx = Math.round(offset.x / 2);
                const dy = Math.round(offset.y / 2);
                this.radiusX = Math.max(1, Math.round(this.editState.radiusX - dx));
                this.radiusY = Math.max(1, Math.round(this.editState.radiusY - dy));
            }
        },
        {
            cursor: 'nesw-resize',
            getRect: (): Rect =>
                new Rect(new Point(this.bounds.x, this.bounds.y + this.bounds.h), new Point(3)).subtract(
                    1.5,
                    1.5,
                    0,
                    0
                ),
            move: (offset: Point): void => {
                const dx = Math.round(offset.x / 2);
                const dy = Math.round(offset.y / 2);
                if (Math.ceil(this.editState.radiusX + dx) >= 1) {
                    this.position.x = this.editState.position.x - dx * 2;
                }
                this.radiusX = Math.max(1, Math.ceil(this.editState.radiusX + dx));
                this.radiusY = Math.max(1, Math.ceil(this.editState.radiusY - dy));
            }
        },
        {
            cursor: 'nwse-resize',
            getRect: (): Rect =>
                new Rect(new Point(this.bounds.x, this.bounds.y), new Point(3)).subtract(1.5, 1.5, 0, 0),
            move: (offset: Point): void => {
                const dx = Math.round(offset.x / 2);
                const dy = Math.round(offset.y / 2);

                if (Math.ceil(this.editState.radiusX + dx) >= 1) {
                    this.position.x = this.editState.position.x - dx * 2;
                }
                if (Math.ceil(this.editState.radiusY + dy) >= 1) {
                    this.position.y = this.editState.position.y - dy * 2;
                }
                this.radiusX = Math.max(1, Math.ceil(this.editState.radiusX + dx));
                this.radiusY = Math.max(1, Math.ceil(this.editState.radiusY + dy));
            }
        }
    ];

    startEdit(mode: EditMode, point: Point, editPoint: TLayerEditPoint) {
        this.mode = mode;
        if (mode == EditMode.CREATING) {
            this.position = point.clone();
            this.radiusX = 1;
            this.radiusY = 1;
            this.updateBounds();
            this.draw();
        }
        this.editState = {
            firstPoint: point,
            position: this.position.clone(),
            radiusX: this.radiusX,
            radiusY: this.radiusY,
            editPoint
        };
    }

    edit(point: Point, originalEvent: MouseEvent) {
        const {position, firstPoint, editPoint} = this.editState;
        switch (this.mode) {
            case EditMode.MOVING:
                this.position = position.clone().add(point.clone().subtract(firstPoint)).round();
                break;
            case EditMode.RESIZING:
                // todo
                editPoint.move(firstPoint.clone().subtract(point));
                break;
            case EditMode.CREATING:
                const radius = point.clone().subtract(firstPoint).abs().divide(2).round().subtract(2);
                this.radiusX = Math.max(radius.x, 0);
                this.radiusY = Math.max(radius.y, 0);
                const signs = point.clone().subtract(firstPoint).xy.map(Math.sign);
                this.position = firstPoint.min(
                    firstPoint
                        .clone()
                        .add(new Point(radius.clone().multiply(2).add(1)).multiply(signs))
                        .floor()
                );
                break;
        }
        this.updateBounds();
        this.draw();
    }

    stopEdit() {
        this.mode = EditMode.NONE;
        this.editState = null;
        this.saveState();
        this.history.push(this.state);
    }

    draw() {
        const {dc, radiusX, radiusY, position} = this;
        const center = position.clone().add(radiusX, radiusY);
        dc.clear();
        dc.ctx.fillStyle = this.color;
        dc.ctx.strokeStyle = this.color;
        dc.pixelateEllipse(center, radiusX, radiusY, this.fill);
    }

    saveState() {
        const state: TEllipseState = {
            p: this.position.xy,
            rx: this.radiusX,
            ry: this.radiusY,
            f: this.fill,
            n: this.name,
            i: this.index,
            g: this.group,
            t: this.type,
            u: this.uid,
            c: this.color,
            in: this.inverted
        };
        this.state = state;
    }

    loadState(state: TEllipseState) {
        this.position = new Point(state.p);
        this.radiusX = state.rx;
        this.radiusY = state.ry;
        this.name = state.n;
        this.index = state.i;
        this.group = state.g;
        this.uid = state.u;
        this.color = state.c;
        this.fill = state.f;
        this.inverted = state.in;
        this.updateBounds();
        this.mode = EditMode.NONE;
    }

    updateBounds() {
        this.bounds = new Rect(this.position, new Point(this.radiusX, this.radiusY).multiply(2)).add(0, 0, 1, 1);
    }
}
