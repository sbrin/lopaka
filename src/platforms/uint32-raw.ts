import {AbstractLayer} from '../core/layers/abstract.layer';
import {bdfFonts} from '../draw/fonts/fontTypes';
import {imgDataToUint32Array} from '../utils';
import {Platform} from './platform';
export class Uint32RawPlatform extends Platform {
    public static id = 'uint32';
    protected name = 'Uint32 Bitmap';
    protected description = 'Uint32 Bitmap';

    protected fonts: TPlatformFont[] = [...bdfFonts];
    constructor() {
        super();
        this.features.hasInvertedColors = true;
        this.features.defaultColor = '#FFFFFF';
    }

    public generateSourceCode(layers: AbstractLayer[], ctx: OffscreenCanvasRenderingContext2D): string {
        const imageData = ctx.getImageData(0, 0, ctx.canvas.width, ctx.canvas.height);
        const UINT32 = imgDataToUint32Array(imageData);
        const iconName = `image_frame`;
        return `const uint32_t ${iconName}[] = {${UINT32}};`;
    }

    public getTemplates() {
        return {};
    }

    public getTemplate() {
        return '';
    }
}
