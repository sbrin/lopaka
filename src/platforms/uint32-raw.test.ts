import {describe, expect, it} from 'vitest';
import {layersMock} from './layers.mock';
import {Uint32RawPlatform} from './uint32-raw';

describe('Flipper zero platform', () => {
    it('generating source code', () => {
        const platform = new Uint32RawPlatform();
        expect(
            platform.generateSourceCode(layersMock, new OffscreenCanvas(128, 64).getContext('2d'))
        ).toMatchSnapshot();
    });
});
