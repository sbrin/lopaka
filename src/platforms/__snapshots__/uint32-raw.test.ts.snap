// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Flipper zero platform > generating source code 1`] = `"const uint32_t image_frame[] = {0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0};"`;
